### ISSUE TITLE
ShutdownAPI doesn't remove OpenSSL callbacks

### ISSUE THREAD

**Issue by jnoxon** [aws/aws-sdk-cpp#397]
For some reason, cleanup_static_state is never invoked after ShutdownAPI is called. Any application that calls ShutdownAPI and then continues to use OpenSSL will crash in the callbacks.

As a work-around, I have added the following to my app after the SDK shutdown:

```
        CRYPTO_set_id_callback(NULL);
        CRYPTO_set_locking_callback(NULL);
```

**Comment by JonathanHenson** [comment]
you mean this?
https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp#L63

What version are you using? A few months back I realized that when we added AES KeyWrap, we forgot to clean it up. That has been fixed though.

**Comment by jnoxon** [comment]
I'm using .39.

From what I see, nothing ever calls cleanup_static_state. The flag s_InitCleanupOpenSSLFlag is always false.

**Comment by JonathanHenson** [comment]
https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/Aws.cpp#L85

https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/utils/crypto/factory/Factories.cpp#L92

https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/utils/crypto/factory/Factories.cpp#L137

https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/utils/crypto/factory/Factories.cpp#L182

https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/utils/crypto/factory/Factories.cpp#L265

https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/utils/crypto/factory/Factories.cpp#L265

https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/utils/crypto/factory/Factories.cpp#L443

https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/utils/crypto/factory/Factories.cpp#L509

https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/utils/crypto/factory/Factories.cpp#L651

**Comment by jnoxon** [comment]
All of those situations will not execute due to the flag being false.

**Comment by JonathanHenson** [comment]
You control the flag

**Comment by JonathanHenson** [comment]
This was updated in 1.0.25

**Comment by jnoxon** [comment]
Facepalm. I somehow missed the method that sets that. I'll give it a shot, thanks.

**Comment by JonathanHenson** [comment]
Per this line:
https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/source/Aws.cpp#L85

It reads from sdk options,  Are you setting SDKOptions::CryptoOptions::initAndCleanupOpenSSL to false?

If you are that's why it isn't being cleaned up. If not, then there is definitely a bug somewhere.

**Comment by jnoxon** [comment]
I'm not using any SDK options at all. My init code is identical to the example here:

https://github.com/awsdocs/aws-doc-sdk-examples/blob/master/cpp/example_code/s3/create_bucket.cpp#L23-L24

**Comment by JonathanHenson** [comment]
You don't need a method to set it. It is read from SDKOptions. That's where this should be configured. If you aren't manually setting this to false, then there is another issue we need to figure out.

That flag is specifically for your use case. You are using OpenSSL elsewhere in your system, and you don't want the AWS SDK changing static global state that the rest of your application may be depending on. 

If that is the case, then I recommend you set this flag (and possibly the curl flag) to false if you want to control the init and shutdown of openssl or curl. If you are content to let us handle that for you (my recommendation) just leave the options as default, move the InitAPI() call above your other init calls to openssl or curl, then delete your manual init calls and use ours.

**Comment by jnoxon** [comment]
Looks like it is set to true by default:

`(gdb) inspect this->SDKOptions.cryptoOptions.initAndCleanupOpenSSL
$5 = true`

**Comment by JonathanHenson** [comment]
It is, and it is never set again so we need to verify.
a.) is it false when you are debugging the shutdown sequence?
   if it is (which I doubt):
     then there is either something, setting it somewhere or there is a corrupt stack somewhere.
  if it isn't (more likely):
    It isn't clearing the callbacks, because the callbacks only get installed iff they aren't already set at the time of init, and we only uninstall them if we successfully set them.

**Comment by jnoxon** [comment]
Here's a gist that demonstrates the issue:

https://gist.github.com/jnoxon/d58162dbcba4e39a4b74ec1fec0b4d7d

```
$ ./test
Nothing initialized yet (expect all DISABLED)
openssl locking callback is DISABLED
openssl id callback is DISABLED
Testing OpenSSL, should not crash...ok

AWS SDK is initialized (expect all ENABLED)
openssl locking callback is ENABLED
openssl id callback is ENABLED
Testing OpenSSL, should not crash...ok

AWS SDK has been shut down (expect all DISABLED)
openssl locking callback is ENABLED
openssl id callback is DISABLED
Segmentation fault
```

**Comment by JonathanHenson** [comment]
We just reworked the shutdown code after discovering a bug with the callback uninstall order. Could you pull latest and try again?

**Comment by singku** [comment]
Fixed in pr #474
